File: EndGate/tests/EndGate.Core.Client.JS.Tests/Tests/UnitTests/BoundingObjects/BoundingRectangleFacts.js
(function (window, bo, assets) {

    QUnit.module("Bounding Rectangle Facts");

    QUnit.test("Un-rotated Rectangle Corners Get works", function () {
        var rect = new bo.BoundingRectangle(6, 6);
        rect.Position = new assets.Vector2d(3, 3);

        QUnit.ok(rect.TopLeft().Equivalent(assets.Vector2d.Zero()));
        QUnit.ok(rect.TopRight().Equivalent(new assets.Vector2d(6, 0)));
        QUnit.ok(rect.BotLeft().Equivalent(new assets.Vector2d(0, 6)));
        QUnit.ok(rect.BotRight().Equivalent(new assets.Vector2d(6, 6)));

        rect.Position = rect.Position.Subtract(6);
        QUnit.ok(rect.TopLeft().Equivalent(new assets.Vector2d(-6, -6)));
        QUnit.ok(rect.TopRight().Equivalent(new assets.Vector2d(0, -6)));
        QUnit.ok(rect.BotLeft().Equivalent(new assets.Vector2d(-6, 0)));
        QUnit.ok(rect.BotRight().Equivalent(assets.Vector2d.Zero()));
    });

    QUnit.test("Rotated Rectangle Corners Get works", function () {
        var rect = new bo.BoundingRectangle(4, 2);
        rect.Position = new assets.Vector2d(3, 2);
        rect.Rotation = Math.PI / 2;

        var topLeft = rect.TopLeft().Clone();
        var topRight = rect.TopRight().Clone();
        var botLeft = rect.BotLeft().Clone();
        var botRight = rect.BotRight().Clone();

        topLeft.Apply(Math.round);
        topRight.Apply(Math.round);
        botLeft.Apply(Math.round);
        botRight.Apply(Math.round);

        QUnit.ok(topLeft.Equivalent(new assets.Vector2d(2, 4)));
        QUnit.ok(topRight.Equivalent(new assets.Vector2d(2, 0)));
        QUnit.ok(botLeft.Equivalent(new assets.Vector2d(4, 4)));
        QUnit.ok(botRight.Equivalent(new assets.Vector2d(4, 0)));
    });

    QUnit.test("Un-rotated Rectangle Corners Get works", function () {
        var rect1 = new bo.BoundingRectangle(10, 6);
        rect1.Position = new assets.Vector2d(5, 3);
        
        var rect2 = new bo.BoundingRectangle(3, 3);
        rect2.Position = new assets.Vector2d(5, 3);

        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));

        rect2.Position = rect2.Position.Add(3);
        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));

        rect2.Position = rect2.Position.Add(1);
        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));

        rect2.Position = rect2.Position.Add(1);
        QUnit.ok(!rect1.Intersects(rect2));
        QUnit.ok(!rect2.Intersects(rect1));
    });

    QUnit.test("Rotated Rectangles collide correctly", function () {
        var rect1 = new bo.BoundingRectangle(4.24, 2.83);
        rect1.Position = new assets.Vector2d(2.5, 2.5);
        rect1.Rotation = -Math.PI / 4;

        var rect2 = new bo.BoundingRectangle(6, 4);
        rect2.Position = new assets.Vector2d(9, 4);

        QUnit.ok(!rect1.Intersects(rect2));
        QUnit.ok(!rect2.Intersects(rect1));

        rect2.Position.X -= 2;

        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));

        rect2.Rotation = Math.PI / 2;
        rect2.Position = rect2.Position.Add(1);

        QUnit.ok(!rect1.Intersects(rect2));
        QUnit.ok(!rect2.Intersects(rect1));

        rect2.Rotation = Math.PI;

        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));

        rect2.Position.Y += 10;

        QUnit.ok(!rect1.Intersects(rect2));
        QUnit.ok(!rect2.Intersects(rect1));

        rect2.Position.Y -= 10;
        rect1.Position.X += 5;

        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));

        rect1.Position.X += 9;

        QUnit.ok(!rect1.Intersects(rect2));
        QUnit.ok(!rect2.Intersects(rect1));

        rect1.Rotation = 0;
        rect1.Position.X = 12;
        rect2.Rotation = Math.PI;

        QUnit.ok(rect1.Intersects(rect2));
        QUnit.ok(rect2.Intersects(rect1));
    });

    QUnit.test("Un-rotated rectangle contains point works.", function () {
        var rect = new bo.BoundingRectangle(6, 4);        
        var vertices;

        rect.Position = new assets.Vector2d(3, 2);

        vertices = rect.Vertices();

        QUnit.ok(rect.ContainsPoint(rect.Position));

        for (var i = 0; i < vertices.length; i++)
        {
            var vertex = vertices[i];
            QUnit.ok(rect.ContainsPoint(vertex));
        }

        QUnit.ok(!rect.ContainsPoint(new assets.Vector2d(-1, 0)));
        QUnit.ok(!rect.ContainsPoint(rect.TopRight().Add(1)));
        QUnit.ok(!rect.ContainsPoint(rect.BotRight().Add(1)));
        QUnit.ok(!rect.ContainsPoint(rect.TopLeft().Subtract(1)));
        QUnit.ok(!rect.ContainsPoint(rect.BotLeft().Subtract(1)));
    });

    QUnit.test("Rotated rectangle contains point works.", function () {
        var rect = new bo.BoundingRectangle(4.24, 2.83);
        var vertices;

        rect.Position = new assets.Vector2d(2.5, 2.5);
        rect.Rotation = -Math.PI / 4;

        vertices = rect.Vertices();

        QUnit.ok(rect.ContainsPoint(rect.Position));

        for (var i = 0; i < vertices.length; i++) {
            var vertex = vertices[i];
            QUnit.ok(rect.ContainsPoint(vertex));
        }

        QUnit.ok(!rect.ContainsPoint(new assets.Vector2d(-1, 0)));
        QUnit.ok(!rect.ContainsPoint(rect.TopRight().Add(1)));
        QUnit.ok(!rect.ContainsPoint(rect.BotRight().Add(1)));
        QUnit.ok(!rect.ContainsPoint(rect.TopLeft().Subtract(1)));
        QUnit.ok(!rect.ContainsPoint(rect.BotLeft().Subtract(1)));

        QUnit.ok(rect.ContainsPoint(rect.TopRight().Subtract(1)));
        QUnit.ok(rect.ContainsPoint(rect.BotRight().Subtract(1)));
        QUnit.ok(rect.ContainsPoint(rect.TopLeft().Add(1)));
        QUnit.ok(rect.ContainsPoint(rect.BotLeft().Add(1)));
    });

})(window, EndGate.Core.BoundingObject, EndGate.Core.Assets);

File: EndGate/tests/EndGate.Core.Net.Tests/BoundingObjects/BoundingRectangleFacts.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using EndGate.Core.Net.Assets;
using EndGate.Core.Net.BoundingObject;
using Xunit;

namespace EndGate.Core.Net.Tests
{
    public class BoundingRectangleFacts
    {
        [Fact]
        public void UnrotatedRectangleCornersGetWorks()
        {
            var rect = new BoundingRectangle(6, 6)
            {
                Position = new Vector2d(3, 3)
            };

            Assert.True(rect.TopLeft.Equivalent(Vector2d.Zero));
            Assert.True(rect.TopRight.Equivalent(new Vector2d(6, 0)));
            Assert.True(rect.BotLeft.Equivalent(new Vector2d(0, 6)));
            Assert.True(rect.BotRight.Equivalent(new Vector2d(6, 6)));

            rect.Position -= 6;
            Assert.True(rect.TopLeft.Equivalent(new Vector2d(-6, -6)));
            Assert.True(rect.TopRight.Equivalent(new Vector2d(0, -6)));
            Assert.True(rect.BotLeft.Equivalent(new Vector2d(-6, 0)));
            Assert.True(rect.BotRight.Equivalent(Vector2d.Zero));
        }

        [Fact]
        public void RotatedRectangleCornersGetWorks()
        {
            var rect = new BoundingRectangle(4, 2)
            {
                Position = new Vector2d(3, 2),
                Rotation = -Math.PI / 2
            };

            var topLeft = rect.TopLeft.Clone();
            var topRight = rect.TopRight.Clone();
            var botLeft = rect.BotLeft.Clone();
            var botRight = rect.BotRight.Clone();

            topLeft.Apply(Math.Round);
            topRight.Apply(Math.Round);
            botLeft.Apply(Math.Round);
            botRight.Apply(Math.Round);

            Assert.True(topLeft.Equivalent(new Vector2d(2, 4)));
            Assert.True(topRight.Equivalent(new Vector2d(2, 0)));
            Assert.True(botLeft.Equivalent(new Vector2d(4, 4)));
            Assert.True(botRight.Equivalent(new Vector2d(4, 0)));
        }

        [Fact]
        public void UnRotatedRectanglesCollideCorrectly()
        {
            var rect1 = new BoundingRectangle(10, 6)
            {
                Position = new Vector2d(5, 3)
            };
            var rect2 = new BoundingRectangle(3, 3)
            {
                Position = new Vector2d(5, 3)
            };

            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));

            rect2.Position += 3;
            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));

            rect2.Position++;
            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));

            rect2.Position++;
            Assert.False(rect1.Intersects(rect2));
            Assert.False(rect2.Intersects(rect1));
        }

        [Fact]
        public void RotatedRectanglesCollideCorrectly()
        {
            var rect1 = new BoundingRectangle(4.24, 2.83)
            {
                Position = new Vector2d(2.5, 2.5),
                Rotation = - Math.PI / 4
            };
            var rect2 = new BoundingRectangle(6, 4)
            {
                Position = new Vector2d(9, 4)
            };

            Assert.False(rect1.Intersects(rect2));
            Assert.False(rect2.Intersects(rect1));

            rect2.Position.X-=2;

            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));

            rect2.Rotation = Math.PI / 2;
            rect2.Position++;

            Assert.False(rect1.Intersects(rect2));
            Assert.False(rect2.Intersects(rect1));

            rect2.Rotation = Math.PI;
            rect2.Position--;

            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));

            rect2.Position.Y+=10;

            Assert.False(rect1.Intersects(rect2));
            Assert.False(rect2.Intersects(rect1));

            rect2.Position.Y -= 10;
            rect1.Position.X += 5;

            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));

            rect1.Position.X += 9;

            Assert.False(rect1.Intersects(rect2));
            Assert.False(rect2.Intersects(rect1));

            rect1.Rotation = 0;
            rect1.Position.X = 12;
            rect2.Rotation = Math.PI;

            Assert.True(rect1.Intersects(rect2));
            Assert.True(rect2.Intersects(rect1));
        }

        [Fact]
        public void UnrotatedRectangleContainsPointWorks()
        {
            var rect = new BoundingRectangle(6, 4)
            {
                Position = new Vector2d(3, 2)
            };

            Assert.True(rect.ContainsPoint(rect.Position));

            foreach (var vertex in rect.Vertices)
            {
                Assert.True(rect.ContainsPoint(vertex));
            }

            Assert.False(rect.ContainsPoint(new Vector2d(-1, 0)));
            Assert.False(rect.ContainsPoint(rect.TopRight + 1));
            Assert.False(rect.ContainsPoint(rect.BotRight + 1));
            Assert.False(rect.ContainsPoint(rect.TopLeft - 1));
            Assert.False(rect.ContainsPoint(rect.BotLeft - 1));
        }

        [Fact]
        public void RotatedRectangleContainsPointWorks()
        {
            var rect = new BoundingRectangle(4.24, 2.83)
            {
                Position = new Vector2d(2.5, 2.5),
                Rotation = Math.PI / 4
            };

            Assert.True(rect.ContainsPoint(rect.Position));

            foreach (var vertex in rect.Vertices)
            {
                Assert.True(rect.ContainsPoint(vertex));
            }

            Assert.False(rect.ContainsPoint(new Vector2d(-1, 0)));
            Assert.False(rect.ContainsPoint(rect.TopRight + 1));
            Assert.False(rect.ContainsPoint(rect.BotRight + 1));
            Assert.False(rect.ContainsPoint(rect.TopLeft - 1));
            Assert.False(rect.ContainsPoint(rect.BotLeft - 1));

            Assert.True(rect.ContainsPoint(rect.TopRight - 1));
            Assert.True(rect.ContainsPoint(rect.BotRight - 1));
            Assert.True(rect.ContainsPoint(rect.TopLeft + 1));
            Assert.True(rect.ContainsPoint(rect.BotLeft + 1));
        }
    }
}
